File: files/300-399/374/CH-374.py
import pandas as pd
from itertools import product

path = "300-399/374/CH-374 Text Cleaning.xlsx"
input = pd.read_excel(path, usecols="B", skiprows=2, nrows=6)
test = pd.read_excel(path, usecols="E", skiprows=2, nrows=6).rename(columns=lambda x: x.replace('.1', ''))
test['ID'] = test['ID'].fillna("")

def bounded_substrings(s):
    chars = list(s)
    n = len(chars)
    
    substrings = []
    for i, j in product(range(n - 1), range(1, n)):
        if i < j and chars[i] == chars[j]:
            substrings.append(''.join(chars[i + 1:j]))
    
    return substrings

input['substrings'] = input['ID'].apply(lambda x: ', '.join(bounded_substrings(x)))
print(input['substrings'].equals(test['ID']))
# > True

File: files/300-399/374/CH-374.R
library(tidyverse)
library(readxl)

path <- "300-399/374/CH-374 Text Cleaning.xlsx"
input <- read_excel(path, range = "B3:B8")
test <- read_excel(path, range = "E3:E8") %>%
  replace_na(list(ID = ""))

bounded_substrings <- function(s) {
  chars <- strsplit(s, "")[[1]]
  n <- length(chars)

  expand.grid(i = 1:(n - 1), j = 2:n) |>
    filter(i < j, chars[i] == chars[j]) |>
    mutate(
      val = purrr::map2_chr(
        i,
        j,
        ~ paste(chars[(.x + 1):(.y - 1)], collapse = "")
      )
    ) |>
    pull(val)
}

result = input %>%
  mutate(
    substrings = map(ID, bounded_substrings) %>% map_chr(paste, collapse = ", ")
  )

all.equal(result$substrings, test$ID)
## [1] TRUE
